### PR TITLE
Update osproc.nim to allow passing stdin, stdout and stderr handles i…

### PR DESCRIPTION
**Pull request by chrisheller**
…n the ''startProcess'' proc.  Includes tests to validate this.

This is for issue #6001

### CODE REVIEW

File: lib/pure/osproc.nim [hunk 1]
```diff
@@ -367,6 +378,11 @@ when not defined(useNimRtl):
       elif not running(p): break
     close(p)
 
+proc assertNoHandlesWithParentStreams(hIn, hOut, hErr: Handle) = 
+  let msg = "Can not pass in/out/err handles when poParentStreams is set"
```
Comment (Araq) [inline lib/pure/osproc.nim]: Make that a 'const' please and the 'proc' a 'template'.

File: tests/osproc/tstderr_fh.nim [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+--------------------------------------'''
+"""
+
+# test that stderr filehandle can be supplied directly
+
+import osproc, os, streams
+when defined(windows):
+  from winlean import get_osfhandle, Handle
+
+const filename = when defined(Windows): "ta_out.exe" else: "ta_out"
```
Comment (Araq) [inline tests/osproc/tstderr_fh.nim]: const filename = addFileExt("ta_out", ExeExt)

File: lib/pure/osproc.nim [hunk 3]
```diff
@@ -517,18 +556,28 @@ when defined(Windows) and not defined(useNimRtl):
       si.dwFlags = STARTF_USESTDHANDLES # STARTF_USESHOWWINDOW or
       if poInteractive notin options:
         createPipeHandles(si.hStdInput, hi)
+        if hIn != INVALID_HANDLE_VALUE:
+          si.hStdInput = hIn
```
Comment (Araq) [inline lib/pure/osproc.nim]: Is si.hStdInput ever closed?

File: lib/pure/osproc.nim [hunk 4]
```diff
@@ -517,18 +556,28 @@ when defined(Windows) and not defined(useNimRtl):
...
+
         createPipeHandles(ho, si.hStdOutput)
+        if hOut != INVALID_HANDLE_VALUE:
+          si.hStdOutput = hOut
```
Comment (Araq) [inline lib/pure/osproc.nim]: Is si.hStdOutput ever closed?

File: lib/pure/osproc.nim [hunk 5]
```diff
@@ -517,18 +556,28 @@ when defined(Windows) and not defined(useNimRtl):
...
+
         if poStdErrToStdOut in options:
           si.hStdError = si.hStdOutput
           he = ho
         else:
           createPipeHandles(he, si.hStdError)
+          if hErr != INVALID_HANDLE_VALUE:
+            si.hStdError = hErr
```
Comment (Araq) [inline lib/pure/osproc.nim]: Is si.hStdError ever closed?

File: lib/pure/osproc.nim [hunk 6]
```diff
@@ -759,6 +811,21 @@ elif not defined(useNimRtl):
          pipe(pStderr) != 0'i32:
         raiseOSError(osLastError())
 
+      if hIn != INVALID_HANDLE_VALUE:
+        discard close(pStdin[readIdx])
+        pStdin[readIdx] = hIn
```
Comment (Araq) [inline lib/pure/osproc.nim]: And these might leak too.

File: lib/pure/osproc.nim (hunk 3 above)
Comment (chrisheller) [inline lib/pure/osproc.nim]: Good catch.  They are closed before overwriting in the non-Windows code path, but not here. I will update these.

File: lib/pure/osproc.nim (hunk 6 above)
Comment (chrisheller) [inline lib/pure/osproc.nim]: pStdin[readIdx], pStdout[writeIdx] and pStderr[writeIdx] are closed by the original code at the end of the proc so those shouldn't leak.

**Comment by Araq** [comment]
Sorry this design has an inherent problem: How can one pass a pipe as the handles? Pipes have to 2 file descriptors that need to be closed in an asymmetric fashion. I've been thinking of extending Nim's ``Stream`` with a ``clientClose`` proc that's usually a nop but wouldn't for pipes that implement the Stream interface and osproc would be aware of clientClose and call it. Then the API would look like ``stdin, stdout, stderr: Stream = nil`` and would be easier to work with, I hope.

**Comment by chrisheller** [comment]
Yes,  I think that sounds like a good plan.

**Comment by dom96** [comment]
I'm going to close this and let the issue https://github.com/nim-lang/Nim/issues/6001 track this.

**Comment by timotheecour** [comment]
> Sorry this design has an inherent problem: How can one pass a pipe as the handles?

Should be fine, this is what's done in https://dlang.org/phobos/std_process.html
pipeProcess (and all other functions) are implemented in terms of spawnProcess, see in code: https://github.com/dlang/phobos/blob/224a1f83ee36e347ac211a6b65502c58f9207959/std/process.d#L2148

**Pull request closed without merging**